### PR TITLE
Export mesh contours & resampling

### LINKED ISSUE #31550: Ability to export contours
Add a support in core renderer to be able to create a smooth contour based on the mesh layer triangulation. Contours should be possible to export by processing algorithm to vector layer.

Contours generation should not be based on GDAL, but rather on custom detection based on the triangles in derived mesh

### PR DESCRIPTION
**Pull request by PeterPetrik**
fixes https://github.com/qgis/QGIS/issues/31550

added the new option in the mesh layer styling dialog to be able to resample face-defined data to vertices. This was required for contour export algorithm, since it is well defined only on the data defined on vertices. 

Screenshot 2019-10-11 at 10 39 16

Screenshot 2019-10-11 at 10 39 23

Also added new exporter class to analysis library for exporting contour polygons and lines. The new class is used in Crayfish plugin in new processing algorithm for export (PR in Crayfish will be soon linked). There is a plan to port the algorithm to QGIS processing core algs later.

Screenshot 2019-10-11 at 10 44 43

and result

Screenshot 2019-10-10 at 17 50 42

Screenshot 2019-10-10 at 17 51 20

### CODE REVIEW

File: src/analysis/mesh/qgsmeshcontours.cpp [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+      ( min_value <= values[2] ) &&( max_value >= values[2] )
+    };
+
+
+    // all values are inside the range == take whole triangle
+    if ( valueInRange[0] && valueInRange[1] && valueInRange[2] )
+    {
+      QVector<QgsPointXY> ring = coords;
+      ring.push_back( coords[0] );
+      const QgsPolygonXY poly = {ring};
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcontours.cpp]: Shouldn't the polygons here be z or m dimensioned?

File: src/analysis/mesh/qgsmeshcontours.cpp [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        }
+      }
+    }
+
+    // add if the polygon is not degraded
+    if ( ring.size() > 2 )
+    {
+      ring.push_back( ring[0] );
+      const QgsPolygonXY poly = {ring};
+      multiPolygon.push_back( QgsGeometry::fromPolygonXY( poly ) );
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcontours.cpp]: This method is very inefficient. See https://github.com/qgis/QGIS/commit/e70f11c4e1bbdc6b99165c35014a3112e95e0939 for the most efficient method for creating new polygon geometries.

File: src/analysis/mesh/qgsmeshcontours.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    ~QgsMeshContours();
+
+    /**
+     * Exports multi line string containing the contour line for particular dataset and value
+     * \param index dataset index
+     * \param value value of the contour line
+     * \param method for datasets defined on faces, the method will be used to convert data to vertices
+     * \returns MultiLineString geometry containing contour lines
+     */
+    QgsGeometry exportLines( const QgsMeshDatasetIndex &index,
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcontours.h]: Would be good to add qgsfeedback arguments here to allow early cancellation

Comment (PeterPetrik) [reply src/analysis/mesh/qgsmeshcontours.h]: done

File: src/analysis/mesh/qgsmeshcontours.cpp (hunk 2 above)
Comment (PeterPetrik) [reply src/analysis/mesh/qgsmeshcontours.cpp]: like the modified?

**Comment by nyalldawson** [comment]
@wonder-sk did you want to review this one before merge?

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**